src/main/ank/2.2-limits-and-colimits.md

```diff
@@ -0,0 +1,208 @@
+val `p'` = p compose m
+val `q'` = q compose m
+```
+................
+```Haskell
+contramap :: (c' -> c) -> (c -> LimD) -> (c' -> LimD)
+contramap f u = u . f
+```
+```kotlin:ank:silent
+import arrow.core.compose
+object LimD
+
+fun <A, B> contramap(f: (A) -> B, u: (B) -> LimD): ((A) -> LimD) = u compose f
+```
+................
+```Haskell
+f :: a -> b
+g :: a -> b
+```
+```kotlin
+val f: (A) -> B
+val g: (A) -> B
+```
+................
+```Haskell
+p :: c -> a
+q :: c -> b
+```
+```kotlin
+val p: (C) -> A
+val q: (C) -> B
+```
+................
+```Haskell
+q = f . p
+q = g . p
+```
+```kotlin
+val q = f compose p
+val q = g compose p
+```
+................
+```Haskell
+f . p = g . p
+```
+```kotlin
+f compose p == g compose p
+```
+................
+```Haskell
+f (x, y) = 2 * y + x
+g (x, y) = y - x
+```
+```kotlin:ank:playground
+val f = { x: Int, y: Int -> 2 * y + x }
+val g = { x: Int, y: Int -> y - x }
+```
+................
+```Haskell
+p t = (t, (-2) * t)
+```
+```kotlin:ank:playground
+import arrow.core.toT
+
+val p = { t: Int -> t toT (-2) * t }
+```
+................
+```Haskell
+f . p' = g . p'
+```
+```kotlin
+f compose `p'` == g compose `p'`
+```
+................
+```Haskell
+p'() = (0, 0)
+```
+```kotlin:ank:playground
+import arrow.core.toT
+
+fun `p'`() = 0 toT 0
+```
+................
+```Haskell
+h () = 0
+```
+```kotlin:ank:plaground
+fun h() = 0
+```
+................
+```Haskell
+f :: a -> b
+g :: c -> b
+```
+```kotlin
+val f: (A) -> B
+val g: (C) -> B
+```
+................
+```Haskell
+p :: d -> a
+q :: d -> c
+r :: d -> b
+```
+```kotlin
+val p: (D) -> A
+val q: (D) -> C
+val r: (D) -> B
+```
+................
+```Haskell
+g . q = f . q
+```
+```kotlin
+g compose q == f compose q
+```
+................
+```Haskell
+f x = 1.23
+```
+```kotlin
+val f: (A) -> Double = { _: A -> 1.23 }
+```
+................
+```Haskell
+twice f x = f (f x)
+```
+```kotlin
+fun <A> twice(): ((((A) -> A), A) -> A) =
+  { f: ((A) -> A), x: A -> f(f(x)) }
+```
+................
+```Haskell
+f       :: t0
+x       :: t1
+f x     :: t2
+f (f x) :: t3
+```
+```kotlin
+f      : t0
+x      : t1
+f x    : t2
+f (f x): t3
+```
+................
+```Haskell
+twice :: t0 -> t1 -> t3
+```
+```kotlin
+val twice: (t0) -> (t1) -> t3
+```
+................
+```Haskell
+t0 = t1 -> t2 -- because f is applied to x
+t0 = t2 -> t3 -- because f is applied to (f x)
+```
+```kotlin
+t0 = t1 -> t2 // because f is applied to x
+t0 = t2 -> t3 // because f is applied to (f x)
+```
+................
+```Haskell
+t1 = t2 = t3 = Int
+twice :: (Int -> Int) -> Int -> Int
+```
+```kotlin
+fun twice(): ((Int) -> Int, Int) -> Int
+```
+................
+```Haskell
+twice :: (t -> t) -> t -> t
+```
+```kotlin
+fun <A> twice(): ((A) -> A, A) -> A
+```
+................
+```Haskell
+newtype ToString a = ToString (a -> String)
+instance Contravariant ToString where 
+    contramap f (ToString g) = ToString (g . f)
+```
+```kotlin:ank:playground
+import arrow.typeclasses.Show
```

Comment:
ToString here isn't a typeclass, it's a datatype.

The equivalent is: `data class ToString<A>(val f: (A) -> String)`